src/main/java/com/hierynomus/smbj/transport/tcp/AsyncDirectTcpTransport.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C)2016 - SMBJ Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hierynomus.smbj.transport.tcp;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.AsynchronousChannelGroup;
+import java.nio.channels.AsynchronousSocketChannel;
+import java.nio.channels.CompletionHandler;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.hierynomus.protocol.Packet;
+import com.hierynomus.protocol.commons.buffer.Buffer;
+import com.hierynomus.protocol.commons.buffer.Buffer.BufferException;
+import com.hierynomus.smbj.lock.CrossThreadLock;
+import com.hierynomus.smbj.transport.PacketHandlers;
+import com.hierynomus.smbj.transport.TransportException;
+import com.hierynomus.smbj.transport.TransportLayer;
+
+/**
+ * A transport layer over Direct TCP/IP that uses asynchronous I/O.
+ */
+public class AsyncDirectTcpTransport<P extends Packet<P, ?>> implements TransportLayer<P> {
+    public static final int DEFAULT_CONNECT_TIMEOUT = 5000;
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+    private final PacketHandlers<P> handlers;
+    private final AsynchronousSocketChannel socketChannel;
+    private final AsyncPacketReader<P> packetReader;
+    private int soTimeout = 0;
+
+    // AsynchronousSocketChannel doesn't support concurrent writes
```

Comment:
I think it's a limitation, or rather a design choice on their part to build a simple AsynchronousSocketChannel and leave it to their users to worry about concurrent writes. Their JavaDoc states that concurrent writes will result in exceptions.

A possible use case where this will be a problem in smbj is where multiple sessions do concurrent writes on the same connection.
Possible design options for handling this situation in smbj are:
1. Not to address this and leave it up to the smbj user to synchronize their writes. This would require users to somehow release locks after the future completes, so they would also need to tackle cross-thread lock and release.
2. Introduce a queue of pending writes inside AsyncDirectTcpTransport - this is probably the most performant solution (no locking while waiting on I/O) but introduces a bit more implementation complexity and additional state per connection.
3. Use a lock for pending writes inside AsyncDirectTcpTransport - this is inline with a similar solution already present in DirectTcpTransport.

I think option 1 is less preferred, and don't have a strong opinion between 2 and 3. Went with 3 just based on the pattern already in use in DirectTcpTransport.